contracts/volt/VoltV2.sol

```diff
@@ -0,0 +1,440 @@
+        );
+        address signatory = ecrecover(digest, v, r, s);
+        require(signatory != address(0), "Volt: invalid signature");
+        require(signatory == owner, "Volt: unauthorized");
+        require(block.timestamp <= deadline, "Volt: signature expired");
+
+        allowances[owner][spender] = amount;
+
+        emit Approval(owner, spender, amount);
+    }
+
+    /// @notice Get the number of tokens held by the `account`
+    /// @param account The address of the account to get the balance of
+    /// @return The number of tokens held
+    function balanceOf(address account) external view returns (uint256) {
+        return balances[account];
+    }
+
+    /// @notice Transfer `amount` tokens from `msg.sender` to `dst`
+    /// @param dst The address of the destination account
+    /// @param amount The number of tokens to transfer
+    /// @return Whether or not the transfer succeeded
+    function transfer(address dst, uint256 amount) external returns (bool) {
+        _transferTokens(msg.sender, dst, amount);
+        return true;
+    }
+
+    /// @notice Transfer `amount` tokens from `src` to `dst`
+    /// @param src The address of the source account
+    /// @param dst The address of the destination account
+    /// @param amount The number of tokens to transfer
+    /// @return Whether or not the transfer succeeded
+    function transferFrom(
+        address src,
+        address dst,
+        uint256 amount
+    ) external returns (bool) {
+        _spendAllowance(src, amount);
+        _transferTokens(src, dst, amount);
+        return true;
+    }
+
+    /// @notice Delegate votes from `msg.sender` to `delegatee`
+    /// @param delegatee The address to delegate votes to
+    function delegate(address delegatee) public {
+        return _delegate(msg.sender, delegatee);
+    }
+
+    /// @notice Delegates votes from signatory to `delegatee`
+    /// @param delegatee The address to delegate votes to
+    /// @param nonce The contract state required to match the signature
+    /// @param expiry The time at which to expire the signature
+    /// @param v The recovery byte of the signature
+    /// @param r Half of the ECDSA signature pair
+    /// @param s Half of the ECDSA signature pair
+    function delegateBySig(
+        address delegatee,
+        uint256 nonce,
+        uint256 expiry,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public {
+        bytes32 domainSeparator = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                keccak256(bytes(name)),
+                block.chainid,
+                address(this)
+            )
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(DELEGATION_TYPEHASH, delegatee, nonce, expiry)
+        );
+        bytes32 digest = keccak256(
+            abi.encodePacked("\x19\x01", domainSeparator, structHash)
+        );
+        address signatory = ecrecover(digest, v, r, s);
```

Comment:
use ecdsa open zeppelin library

Reply:
agreed we should do this to be fully secure

Reply:
https://medium.com/immunefi/intro-to-cryptography-and-signatures-in-ethereum-2025b6a4a33d#:~:text=Signature%20Malleability,r%2C%20s%2C%20and%20v.